### DISCUSSION TITLE
Memory usage with miller & join

### DISCUSSION THREAD

**Discussion by trantor** [johnkerl/miller#502] [Q&A]
Hello everyone.

I am processing a multi-GB CSV file, attempting to `join` with a different CSV file of just a few lines... and quickly I've got to abort `mlr` since the RSS of the process skyrockets to multi-GB levels as well, quite larger than the original CSV as well.

Much the same things happens when one tries, for instance, a `grep -f ` with dozens of regex patterns (not fixed strings), so I am not exactly surprised. However, since the problem of `join`ing (essentialy to filter and associate down to a relatively small amount of records) on sizeable CSV files is relatively common, I would be curious about options on alternative approaches to the problem.

**Comment by johnkerl** [comment]
Hi @trantor -- you might check out `mlr join -s` -- if the inputs are already sorted on their join key(s).

Another idea is to maybe swap the role of the 'left' and 'right' file. Without `-s`, the 'left' file will be loaded into memory but the 'right' file will be streamed.

**Reply by trantor** [reply]
Thanks @johnkerl, those files are not actually sorted. However the "left" file in the join is precisely the very small one already, so I am not sure your second suggestion helps in this case :disappointed:

**Reply by johnkerl** [reply]
:(

OK. Pre-sorting the big one may help.

Also, what is your Miller version?

**Reply by trantor** [reply]
@johnkerl It's version 5.10.2 .
I've tried pre-sorting both files on the join field, to no avail I am afraid.
Just to give you a better idea, the "right" file is about 2.7 GB in size, while the left file less than 3 KB (31 records in it).
I've killed the `mlr` process once the RSS reached 5 GB (to avoid exhausting the available memory).

**Reply by trantor** [reply]
Mmmm actually it seems that the join might be a red herring in a way.
Below is my actual workflow
```
        mlr --implicit-csv-header --csv \
            rename 1,metadata_path,2,digest then \
            filter '$metadata_path !=~ "\\.del$"' then \
            cat -n -g digest then \
            filter '$n == 1' then \
            cut -x -f n then \
            put '$digest = "backup/items/".substr($digest,0,1)."/" . $digest' then \
            sort -f digest then \
            join -f left_file.csv -s --icsv -l path -r digest -j path \ 
                right_file.csv
```
If I remove all the steps after the first `filter` the RSS memory tops more or less at the filesize of the "right" file.
If I leave any of the following steps the memory consumptions rises swiftly.
I guess that `cat -n -g digest`, in order to perform the grouping and counting correctly, needs to keep in memory more than a few things; however for it to balloon to 2 or 3 times the size of the input data looks strange to me.

**Reply by johnkerl** [reply]
OK -- is `digest` a high-cardinality field?

What does this show:

```
mlr --csv count-distinct -n -f digest right_file.csv
```

?

**Reply by trantor** [reply]
Yes @johnkerl  . 
Not going to actually run the count, but there should be around 18-19 million distinct records

**Reply by johnkerl** [reply]
Hi @trantor I don't mean count of records, I mean count of *distinct* values of `digest` ...

If this is a large number then I can see why `cat -n -g digest` would be taking up more memory ...

**Reply by trantor** [reply]
You are right, and when I said records I intended the number of records with distinct values of `digest`

**Reply by trantor** [reply]
I am not surprised it takes a large amount of memory, I am more surprised about how much more than the entire size of the input file.
And also one thing came to mind. Before the `cat -n -g digest` step, shouldn't the `rename` + `filter` steps be streaming record by record? Why even those steps alone bring the RSS up to around the size of the input file ?

**Reply by trantor** [reply]
Btw @johnkerl  as a sample these are the first 4 lines of the actual input file.
```
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/302","%2BPInrYaKIv0OSIMSwqYk59DeWiODMx%2CQKcOLCRCRGyA%3D"
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/602","CfKi2QeCymoegAPaBS8Pe%2C8fJ0sRwymik%2CHLFzKXiOI%3D"
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/702","cvRdeMKgDe8X0QJWPaBjn%2B8j7xcddQO8INfFRmpGZvU%3D"
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/902","vdtyY8v4Gdx2t2EgK7dka3zuDxq3QifH7ml433jlpMc%3D"
```

**Reply by johnkerl** [reply]
They shouldn't be bringing up the RSS, something is weird.

Reason I asked about your Miller version is I used to use `mmap` and there were memory issues even with streaming large files. But you're in 5.10.

Also the JSON reader for Miller 5 and below is non-streaming (it's streaming in Miller 6) but that's not the issue here either since you have CSV.

Let me see what I can reproduce & get back to you ...

**Reply by trantor** [reply]
@johnkerl Ignore what I said about the memory consumption. I was running the process across several machines and read the version on the wrong one. It was actually version 5.6.2 . With version 5.10.2 the memory consumption with `rename + filter` is actually pretty low. With the `cat -n -g digest` step the RSS tops at 4-4.5 GB but it fluctuates up and down. The following steps do not bring it any higher, so that's probably the "culprit". The size of the input file is around 2.5 GB, as mentioned earlier.

**Reply by johnkerl** [reply]
@trantor excellent, this explains! :) `mmap` was removed in [5.7.0](https://github.com/johnkerl/miller/releases/tag/v5.7.0).

**Reply by johnkerl** [reply]
@trantor meanwhile for 5.6.2, `mlr --no-mmap` should help

**Reply by trantor** [reply]
@johnkerl I am having another strange issue with this workflow... I hope I am missing something glaringly obvious so another pair of eyes might help ;)
I've got a complete "pipeline" returning me no results, when splitting it in two, imho, equivalent steps does. I am somewhat baffled.
```
# Pipeline with no output
mlr --implicit-csv-header --csv \
    rename 1,metadata_path,2,digest then \
    filter '$metadata_path !=~ "\\.del$"' then \
    cat -n -g digest then \
    filter '$n == 1' then \
    cut -x -f n then \
    put '$digest = "backup-mail3/items/".$digest' then \
    sort -f digest then \
    join -f file1.csv -i csv -l path -r digest -j path file2.csv

------

# Intermediate step 1
mlr --implicit-csv-header --csv \
    rename 1,metadata_path,2,digest then \
    filter '$metadata_path !=~ "\\.del$"' then \
    cat -n -g digest then \
    filter '$n == 1' then \
    cut -x -f n then \
    put '$digest = "backup-mail3/items/".$digest' then \
    sort -f digest file2.csv > intermediate1.csv

# Intermediate step 2, returning results
mlr --csv join -f file1.csv  -i csv -l path -r digest -j path intermediate1.csv


# mlr --version
Miller v5.10.2
```
Can you spot a meaningful difference?

**Reply by johnkerl** [reply]
@trantor there is no meaningful difference in what you wrote. :)

I can reproduce using your data above; 👀 ...

**Reply by johnkerl** [reply]
@trantor please add `--no-implicit-csv-header` on your final join

```
$ cat file1.csv
path,foo
backup-mail3/items/%2BPInrYaKIv0OSIMSwqYk59DeWiODMx%2CQKcOLCRCRGyA%3D,1
backup-mail3/items/CfKi2QeCymoegAPaBS8Pe%2C8fJ0sRwymik%2CHLFzKXiOI%3D,2
backup-mail3/items/cvRdeMKgDe8X0QJWPaBjn%2B8j7xcddQO8INfFRmpGZvU%3D,3
backup-mail3/items/vdtyY8v4Gdx2t2EgK7dka3zuDxq3QifH7ml433jlpMc%3D,4
```

```
$ cat file2.csv
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/302","%2BPInrYaKIv0OSIMSwqYk59DeWiODMx%2CQKcOLCRCRGyA%3D"
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/602","CfKi2QeCymoegAPaBS8Pe%2C8fJ0sRwymik%2CHLFzKXiOI%3D"
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/702","cvRdeMKgDe8X0QJWPaBjn%2B8j7xcddQO8INfFRmpGZvU%3D"
"/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/902","vdtyY8v4Gdx2t2EgK7dka3zuDxq3QifH7ml433jlpMc%3D"
```

```
mlr --from file2.csv --implicit-csv-header --csv \
    rename 1,metadata_path,2,digest then \
    filter '$metadata_path !=~ "\\.del$"' then \
    cat -n -g digest then \
    filter '$n == 1' then \
    cut -x -f n then \
    put '$digest = "backup-mail3/items/".$digest' then \
    sort -f digest then \
    join -f file1.csv -i csv -l path -r digest -j path

(nothing)
```

vs

```
mlr --from file2.csv --implicit-csv-header --csv \
    rename 1,metadata_path,2,digest then \
    filter '$metadata_path !=~ "\\.del$"' then \
    cat -n -g digest then \
    filter '$n == 1' then \
    cut -x -f n then \
    put '$digest = "backup-mail3/items/".$digest' then \
    sort -f digest then \
    join -f file1.csv --no-implicit-csv-header -i csv -l path -r digest -j path

path,foo,metadata_path
backup-mail3/items/%2BPInrYaKIv0OSIMSwqYk59DeWiODMx%2CQKcOLCRCRGyA%3D,1,/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/302
backup-mail3/items/CfKi2QeCymoegAPaBS8Pe%2C8fJ0sRwymik%2CHLFzKXiOI%3D,2,/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/602
backup-mail3/items/cvRdeMKgDe8X0QJWPaBjn%2B8j7xcddQO8INfFRmpGZvU%3D,3,/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/702
backup-mail3/items/vdtyY8v4Gdx2t2EgK7dka3zuDxq3QifH7ml433jlpMc%3D,4,/opt/zimbra/backup/zextras/accounts/a0dde7c9-5c48-43e0-b52b-11b0007be7d2/items/02/902
```

since it appears your `file1.csv` has a header, where `file2.csv` does not except you are adding one during processing.

**Reply by aborruso** [reply]
@johnkerl a note: I do not find `--no-implicit-csv-header` in the documentation https://miller.readthedocs.io/en/latest/reference.html

After applying rename verb it would be like having canceled `--implicit-csv-header` option, mlr should inherit it. Am I saying something meaningful and useful?

**Reply by trantor** [reply]
@aborruso I can't find it either...
Anyway @johnkerl thanks for spotting the problem. I guess that since that option is not mentioned in the list for the options related to the left file in  `join` I assumed it would not be inherited.
Can you document those other options as well? Thanks :D

**Reply by johnkerl** [reply]
> mlr should inherit it

@aborruso `mlr join` is inherting `--implicit-csv-header` from `mlr` but `mlr join` is being given data which has a header. So `mlr join` needs to be told `--no-implicit-csv-header`.

> Can you document those other options as well? Thanks :D

Thanks @trantor , will do!
